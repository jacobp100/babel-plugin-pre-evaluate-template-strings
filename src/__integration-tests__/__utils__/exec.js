/* @flow */

import childProcess from 'child_process';
import path from 'path';

export function transpile(
  source: string,
  pluginOptions: any = { cache: false, extract: false },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_TRANSPILE_BIN = path.join(__dirname, './transpile.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_TRANSPILE_BIN} '${JSON.stringify(
      source
    )}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [transpiled, rawStyles] = stdout.toString().split('\n');
  const styles = JSON.parse(rawStyles || '{}');
  return {
    code: JSON.parse(transpiled || '""'),
    styles,
    getCSSForClassName(className: string) {
      return styles[`.${className}`];
    },
  };
}

export function extract(
  source: string,
  pluginOptions: any = { cache: false, extract: true },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_EXTRACT_BIN = path.join(__dirname, './extract.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_EXTRACT_BIN} '${JSON.stringify(source)}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [data, filenames, transpiled] = stdout.toString().split('\n');
  return {
    data: JSON.parse(data || '""'),
    filenames: JSON.parse(filenames || '[]'),
    transpiled: JSON.parse(transpiled || '""'),
  };
}
