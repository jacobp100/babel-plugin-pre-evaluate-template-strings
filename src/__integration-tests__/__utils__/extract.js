const fs = require('fs');

const writeFileSync = fs.writeFileSync;

let buffer = '';
const filenames = [];

fs.writeFileSync = (filename, data, opts) => {
  if (/\.css$/.test(filename)) {
    filenames.push(filename);
    buffer += data;
    return;
  }
  writeFileSync(filename, data, opts);
};

const transpile = require('./transpile');

const code = transpile(
  JSON.parse(process.argv[2]),
  JSON.parse(process.argv[3])
);

console.log(JSON.stringify(buffer));
console.log(JSON.stringify(filenames));
console.log(JSON.stringify(code || ''));
