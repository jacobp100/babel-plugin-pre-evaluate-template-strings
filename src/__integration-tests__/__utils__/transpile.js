const babel = require('babel-core');
const path = require('path');
const dedent = require('dedent');

const CSS_PATH = require.resolve(path.join(process.cwd(), 'build/css'));
const SHEET_PATH = require.resolve(path.join(process.cwd(), 'build/sheet'));
const PREVAL_PLUGIN_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/preval-extract')
);
const MODULE_SYSTEM_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/lib/moduleSystem')
);

function transpile(source, options = { pluginOptions: {}, babelOptions: {} }) {
  return babel.transform(
    dedent`
      import css from '${CSS_PATH}';

      ${source}
    `,
    Object.assign(
      {
        presets: ['env', 'stage-2', 'react'],
        plugins: [[PREVAL_PLUGIN_PATH, options.pluginOptions]],
        sourceMaps: false,
      },
      options.babelOptions
    )
  ).code;
}

module.exports = transpile;

if (require.main && module.filename === require.main.filename) {
  const code = transpile(
    JSON.parse(process.argv[2]),
    JSON.parse(process.argv[3])
  );

  if (code) {
    console.log(JSON.stringify(code));

    // eslint-disable-next-line global-require
    const getCachedModule = require(MODULE_SYSTEM_PATH).getCachedModule;
    const sheet = getCachedModule(SHEET_PATH);
    const styles = sheet
      ? getCachedModule(SHEET_PATH).exports.default.styles()
      : {};
    console.log(JSON.stringify(styles || ''));
  }
}
