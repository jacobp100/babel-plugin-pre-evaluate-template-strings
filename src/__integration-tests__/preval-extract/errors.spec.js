/* eslint-disable no-template-curly-in-string */
/* @flow */

import dedent from 'dedent';

import { transpile } from '../__utils__/exec';

describe('preval-extract babel plugin errors', () => {
  it('should throw error if "css" tagged template literal is not assigned to a variable', () => {
    expect(() => {
      transpile(dedent`
      css\`
        font-size: 3em;
      \`;
      `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw error if "css.named" is not called with classname', () => {
    expect(() => {
      transpile(dedent`
      css.named\`
        font-size: 3em;
      \`;
      `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw error if the id was not found', () => {
    expect(() => {
      transpile(dedent`
      const title = css\`
        width: ${'${document.width}'};
      \`;
      `);
    }).toThrowErrorMatchingSnapshot();
  });
});
