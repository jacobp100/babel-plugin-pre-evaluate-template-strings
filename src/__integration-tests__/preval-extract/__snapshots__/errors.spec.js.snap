// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract babel plugin errors should throw error if "css" tagged template literal is not assigned to a variable 1`] = `
Command failed: node <<CWD>>/src/babel/__integration-tests__/__utils__/transpile.js '"css\`\\n  font-size: 3em;\\n\`;"' '{"pluginOptions":{"cache":false,"extract":false},"babelOptions":{"filename":"test.js"}}'
<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:590
      throw err;
      ^

SyntaxError: test.js: Couldn't determine the class name for CSS template literal. Ensure that it's either:
- Assigned to a variable
- Is an object property
- Is a prop in a JSX element

  1 | import css from '<<CWD>>/build/css.js';
  2 | 
> 3 |     css\`
    |     ^
  4 | font-size: 3em;
  5 | \`;
    at File.buildCodeFrameError (<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:427:15)
    at NodePath.buildCodeFrameError (<<CWD>>/node_modules/babel-traverse/lib/path/index.js:140:26)
    at PluginPass.TaggedTemplateExpression (<<CWD>>/build/babel/preval-extract/index.js:106:24)
    at newFn (<<CWD>>/node_modules/babel-traverse/lib/visitors.js:276:21)
    at NodePath._call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:105:12)
    at TraversalContext.visitQueue (<<CWD>>/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitSingle (<<CWD>>/node_modules/babel-traverse/lib/context.js:108:19)
    at TraversalContext.visit (<<CWD>>/node_modules/babel-traverse/lib/context.js:192:19)

`;

exports[`preval-extract babel plugin errors should throw error if "css.named" is not called with classname 1`] = `
Command failed: node <<CWD>>/src/babel/__integration-tests__/__utils__/transpile.js '"css.named\`\\n  font-size: 3em;\\n\`;"' '{"pluginOptions":{"cache":false,"extract":false},"babelOptions":{"filename":"test.js"}}'
<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:590
      throw err;
      ^

Error: test.js: Linaria's \`css.named\` must be called with a class name
    at isLinariaTaggedTemplate (<<CWD>>/build/babel/preval-extract/validators.js:22:11)
    at PluginPass.TaggedTemplateExpression (<<CWD>>/build/babel/preval-extract/index.js:88:72)
    at newFn (<<CWD>>/node_modules/babel-traverse/lib/visitors.js:276:21)
    at NodePath._call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:105:12)
    at TraversalContext.visitQueue (<<CWD>>/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitSingle (<<CWD>>/node_modules/babel-traverse/lib/context.js:108:19)
    at TraversalContext.visit (<<CWD>>/node_modules/babel-traverse/lib/context.js:192:19)
    at Function.traverse.node (<<CWD>>/node_modules/babel-traverse/lib/index.js:114:17)

`;

exports[`preval-extract babel plugin errors should throw error if the id was not found 1`] = `
Command failed: node <<CWD>>/src/babel/__integration-tests__/__utils__/transpile.js '"const title = css\`\\n  width: \${document.width};\\n\`;"' '{"pluginOptions":{"cache":false,"extract":false},"babelOptions":{"filename":"test.js"}}'
<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:590
      throw err;
      ^

SyntaxError: test.js: Linaria css evaluation error:
  Could not find a reference to 'document'.
  This might happen if you used some undeclared variable/function or a browser specific API.

  2 | 
  3 |     const title = css\`
> 4 | width: \${document.width};
    |          ^
  5 | \`;
    at File.buildCodeFrameError (<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:427:15)
    at NodePath.buildCodeFrameError (<<CWD>>/node_modules/babel-traverse/lib/path/index.js:140:26)
    at resolveSource (<<CWD>>/build/babel/preval-extract/resolveSource.js:35:16)
    at Object.Identifier (<<CWD>>/build/babel/preval-extract/index.js:46:48)
    at NodePath._call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (<<CWD>>/node_modules/babel-traverse/lib/path/context.js:105:12)
    at TraversalContext.visitQueue (<<CWD>>/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitSingle (<<CWD>>/node_modules/babel-traverse/lib/context.js:108:19)
    at TraversalContext.visit (<<CWD>>/node_modules/babel-traverse/lib/context.js:192:19)

`;
